modules/javafx.graphics/src/main/java/javafx/scene/layout/AnchorPane.java

```diff
@@ -355,15 +350,15 @@ private double computeChildHeight(Node child, Double topAnchor, Double bottomAnc
             }
 
             if (leftAnchor != null) {
-                x = insets.getLeft() + leftAnchor;
+                x = snappedLeftInset() + leftAnchor;
```

Comment:
Not sure about render scale - it should be done elsewhere.

As for snapping, I think this code might be incorrect - it should honor Region.snapToPixel setting.  All the intermediary computations need to use unsnapped (original) values, then, at the last moment, call proper Region.snap*() which either rounds or not, depending on the `snapToPixel`.

The only exception for this rule (of using snapped values in the intermediary computations) is to ensure some kind of specific alignment, which I believe is not the case here.

Reply:
I'm not sure about this. Other places do the same, and there are lot of places actually which do something like `snappedLeftInsets` + `snappedRightInsets` or `final double w = snapSizeX(getWidth()) - x - snappedRightInset();` or similar. 
I think this was for a reason. 
And I have only oriented myself to the existing code in this case. If you are right, then this was done wrong everywhere.

Btw. I had a look and other container do snap their constraints, e.g. `H/VBox` its spacing, `GridPane` his `H/VGap`. So pretty sure we also should snap the anchor values.

Reply:
I think the idea is this (and it only applies when snapToPixels=true).  We only need to use snapped coordinates/sizes when placing components.  In order to avoid accumulating errors, which can be seen during gradual resizing for example, we should avoid snapping of intermediary results, UNLESS there is a possibility of introducing visual artifacts.

What I mean by this is imagine a container that aligns its children in a table-like layout (similar to https://github.com/andy-goryachev/FxDock/blob/master/doc/CPane.md).  Depending on constraints, there might be a situation where rounding might shift certain children in relation to the vertical guide lines - we don't want that.

So in this case we need to snap the guide lines as an intermediate step (i.e. compute the first guide line position, then compute the next column, snap, next column, etc.)

Whereas in a simple case (one child node), we can simply use snapped*Inset() - we don't have intermediate computations.

In the case of AnchorPane, I think, we should first sum insets and anchor, then snap - because developers might want to use or not to use snapToPixels, and might set a fractional anchor.

Reply:
I initially had the same thought that Andy had, but I want to take a closer look -- especially at what similar layout panes do.

As for the point about doing this only when `snapPixels == true`, yes, but you don't need to worry about that if you call the `snapXXX` methods in Region. All of them check the `snapToPixel` attribute.

Reply:
I looked at the other layout panes and they already do something similar to what this PR does, using the snapped insets for intermediate compuation. I also note that some of the computation is calling superclass methods that do snapping already, and expect it to be added to snapped insets (if there are insets), for example, `computeWidth()`. In this particular case -- `snappedLeftInset() + leftAnchor` -- I don't think `leftAnchor` is snapped, but it probably should be.

Reply:
Yeah I think so too.

Reply:
I also never understood that completely.
Hence I asked at the mailing list about snapping in the past: https://mail.openjdk.org/pipermail/openjfx-dev/2021-September/032166.html
In my opinion it also makes sense to snap the values once. But that would be a huge refactoring an everything is affected. We can not just change this now. Thats why I would stick to the existing solution and refactor that later.
But some clarification here would be helpful here, maybe ask the people who created this (if available)?. And I haven't looked yet, but how does Swing handle this?

Reply:
It's seems little tricky to choose an approach for this fix.

- NOT snapping intermediate values (new approach):
1. It definitely seems correct NOT to snap intermediate values.
2. But currently, code Controls snap the intermediate values,
3. This PR seems to be the first change of snapping in a layout:
=> Will it regress if we have different approach in layout than in control ?
I think No, It won’t regress. @Maran23 Can you please try it with this fix ?
In that case, we can adapt the new way for layout and file a new JBS to refactor existing code.

Reply:
I think I now understand the snapping better, especially after discussion in https://github.com/openjdk/jfx/pull/1111

it looks like we should be operating in terms of final pixels, that is, in snapped coordinates.  this might mean what we should not, generally, mix snapped and unsnapped values:

result = snap(a) + snap(b); // ok

result = snap(snap(a) + b) // also ok

result = snap(snap(a) + b + c) // not ok, should be written as
result = snap(snap(a) + snap(b) + snap(c)) or
result = snap(a) + snap(b) + snap(c)

What do you think?

Reply:
I agree but that is also what I do here, isn't it?

Reply:
here: yes!

Reply:
This looks good to me, the snapping as explained by Andy is how it should be.  A source of error can be still be that floating point calculations can introduce small deviations so sometimes it is needed to resnap after doing calculations that seem harmless (ie. `snappedSize * 10` or `snappedWidth - snappedLeftInset` may need resnapping).

Reply:
Good point. That is something we should think about - this may or may not be a problem with some of the snapping code.